### PR TITLE
Motif 95

### PR DESCRIPTION
**Pull request by HubertHuckevoll**
This patch resolves issue #407
It also reworks the SPUI-code to be a little more modular.

### CODE REVIEW

File: Library/SpecUI/CommonUI/CUtils/copenUtils.asm
Comment (bluewaysw) [inline Library/SpecUI/CommonUI/CUtils/copenUtils.asm:1]: Only white space changes? Please revert the file....

File: Library/SpecUI/CommonUI/CWin/cwinClassCUAS.asm
Comment (bluewaysw) [inline Library/SpecUI/CommonUI/CWin/cwinClassCUAS.asm:1]: Only empty line changes, could we revert?

File: Library/SpecUI/CommonUI/CWin/cwinClassCommonLow.asm
Comment (bluewaysw) [inline Library/SpecUI/CommonUI/CWin/cwinClassCommonLow.asm:1]: No real change here, could we revert?

File: Library/SpecUI/CommonUI/CWin/cwinClassCommonMiddle.asm
Comment (bluewaysw) [inline Library/SpecUI/CommonUI/CWin/cwinClassCommonMiddle.asm:1]: Not a real change here, could we revert?

File: Library/SpecUI/CommonUI/CWin/cwinClassMisc.asm
Comment (bluewaysw) [inline Library/SpecUI/CommonUI/CWin/cwinClassMisc.asm:1]: Not a real change here, could we revert?

File: Library/SpecUI/CommonUI/CWin/cwinFieldInit.asm [hunk 1]
```diff
@@ -512,6 +515,30 @@ OLFieldEnsureExpressMenu	proc	far
...
+	test	ds:[taskBarPrefs], mask TBF_ENABLED	; test if TBF_ENABLED is set
+	pop	ds					; restore ds
+	jz	done					; skip if no taskbar
+
+	push	si
+	mov	si, ds:[di].OLFI_expressMenu		; set *ds:si = menu
+	mov	ax, HINT_FIXED_SIZE			; ax - Vardata type
+	mov	cx, size SpecWidth + size SpecHeight + size word ; cx - Size of extra data. (cx = zero if no extra data).
+	call	ObjVarAddData				; returns: pointer to extra data in ds:bx
+	mov	ds:[bx].SSA_width, SpecWidth <SST_PIXELS, 30>
```
Comment (bluewaysw) [inline Library/SpecUI/CommonUI/CWin/cwinFieldInit.asm:536]: Is giving a fixed size a good choice here? Think about different screen resolutions and express menu/taskbar sizes....

File: Library/SpecUI/CommonUI/CWin/cwinUtils.asm
Comment (bluewaysw) [inline Library/SpecUI/CommonUI/CWin/cwinUtils.asm:1]: No real change here, could we revert?

File: Library/SpecUI/CommonUI/CWin/cwinFieldUncommon.asm [hunk 2]
```diff
@@ -1832,26 +1834,111 @@ TaskBarListAddChild	method dynamic TaskBarListClass,
...
+					;bx, VI_bounds.R_top
+
+	push	ax
+	call	GetDarkColor		; al <- dark color
+	mov	ah, C_WHITE		; ah <- light color
+	xchg	ah, al			; swap colors, as we want an "outset" look
+	mov	bp, ax
+	pop	ax
+
+; VisBounds ist die Position des Vis-Object innerhalb des Parent
```
Comment (bluewaysw) [inline Library/SpecUI/CommonUI/CWin/cwinFieldUncommon.asm:1923]: Better in english here...

File: Library/SpecUI/CommonUI/CWin/cwinFieldUncommon.asm [hunk 3]
```diff
@@ -2284,37 +2467,109 @@ if TOOL_AREA_IS_TASK_BAR
...
+	; set TaskBar usable
+	;
+	push 	ds, si
+	mov	ax, MSG_GEN_FIND_CHILD_AT_POSITION
+	clr	cx					; find first child = 0
+	call	ObjCallInstanceNoLock
+	jc	afterTaskBarCheck			; abort if child not found
+	mov 	si, dx					; *ds:si <- TaskBar
+	mov	ax, MSG_GEN_SET_USABLE
+	mov	dl, VUM_DELAYED_VIA_APP_QUEUE		;VUM_NOW ;VUM_DELAYED_VIA_APP_QUEUE ; dl <- VisUpdateMode
```
Comment (bluewaysw) [inline Library/SpecUI/CommonUI/CWin/cwinFieldUncommon.asm:2526]: Comment confuses

File: Library/SpecUI/CommonUI/CWin/cwinFieldUncommon.asm [hunk 4]
```diff
@@ -2284,37 +2467,109 @@ if TOOL_AREA_IS_TASK_BAR
...
+	; set Tray usable
+	;
+	push 	ds, si
+	mov	ax, MSG_GEN_FIND_CHILD_AT_POSITION
+	mov	cx, 1					; second child = 1
+	call	ObjCallInstanceNoLock
+	jc	afterTrayCheck				; abort if child not found
+	mov 	si, dx					; *ds:si <- TaskBar
+	mov	ax, MSG_GEN_SET_USABLE
+	mov	dl, VUM_DELAYED_VIA_APP_QUEUE		; VUM_NOW ;VUM_DELAYED_VIA_APP_QUEUE ; dl <- VisUpdateMode
```
Comment (bluewaysw) [inline Library/SpecUI/CommonUI/CWin/cwinFieldUncommon.asm:2541]: Comment confuses...

File: Library/SpecUI/CommonUI/CWin/cwinBase.asm [hunk 5]
```diff
@@ -1400,13 +1399,13 @@ continue:
 	mov	ax, HINT_AVOID_MENU_BAR
 	call	ObjVarFindData
 	jc	done
-if EXTENDIBLE_SYSTEM_TRAY
+;if EXTENDIBLE_SYSTEM_TRAY ; - this flag is dead. Taskbar also means Systray.
```
Comment (bluewaysw) [inline Library/SpecUI/CommonUI/CWin/cwinBase.asm:1402]: Please cleanup the condition here....

File: Library/User/UI/uiExpress.asm [hunk 6]
```diff
@@ -642,9 +642,16 @@ afterMainAppsList:
 	mov	ax, MSG_GEN_SET_NOT_USABLE
 	mov	dl, VUM_DELAYED_VIA_UI_QUEUE
 	call	ObjCallInstanceNoLock
+
 	mov	ax, MSG_GEN_INTERACTION_SET_VISIBILITY
 	mov	cl, GIV_POPUP
 	call	ObjCallInstanceNoLock
+
+	mov	cx, offset OtherAppsListMoniker
```
[the remark this answers is not in the document]
Comment (HubertHuckevoll) [reply Library/User/UI/uiExpress.asm:650]: We've talked about this here: https://bluewaysw.slack.com/archives/DRG0189KM/p1706399554345749

File: Library/User/UI/uiExpress.ui
[the remark this answers is not in the document]
Comment (HubertHuckevoll) [reply Library/User/UI/uiExpress.ui:1]: Okay, I will revert this.

File: Include/Internal/SpecUI/olWinClass.def
[the remark this answers is not in the document]
Comment (HubertHuckevoll) [reply Include/Internal/SpecUI/olWinClass.def:1]: Okay, I will revert this, if it was only whites-space changes.

File: Include/Internal/SpecUI/olApplicationClass.def
[the remark this answers is not in the document]
Comment (HubertHuckevoll) [reply Include/Internal/SpecUI/olApplicationClass.def:1]: Okay, I will revert this.

File: Library/SpecUI/CommonUI/CSpec/cspecCUAS.ui [hunk 7]
```diff
@@ -1168,59 +1156,76 @@ FloatingToolArea = ToolArea {
 		ATTR_GEN_WINDOW_CUSTOM_WINDOW_PRIORITY {
 		    word WIN_PRIO_STD-1
 		},
-#endif
-#if !_ISUI
+#else
+#if _MOTIF
+		HINT_STAGGER_WINDOW,
```
[the remark this answers is not in the document]
Comment (HubertHuckevoll) [reply Library/SpecUI/CommonUI/CSpec/cspecCUAS.ui:1161]: I will look this up again...

File: Library/SpecUI/CommonUI/CUtils/copenUtils.asm
Comment (HubertHuckevoll) [reply Library/SpecUI/CommonUI/CUtils/copenUtils.asm:1]: Okay, I will revert this if it is only white-space changes.

File: Library/SpecUI/CommonUI/CWin/cwinClassCUAS.asm
Comment (HubertHuckevoll) [reply Library/SpecUI/CommonUI/CWin/cwinClassCUAS.asm:1]: Okay, I will revert this if it is only white-space changes.

File: Library/SpecUI/CommonUI/CWin/cwinClassCommonLow.asm
Comment (HubertHuckevoll) [reply Library/SpecUI/CommonUI/CWin/cwinClassCommonLow.asm:1]: Okay, I will revert this if it is only white-space changes.

File: Library/SpecUI/CommonUI/CWin/cwinClassCommonMiddle.asm
Comment (HubertHuckevoll) [reply Library/SpecUI/CommonUI/CWin/cwinClassCommonMiddle.asm:1]: Okay, I will revert this if it is only white-space changes.

File: Library/SpecUI/CommonUI/CWin/cwinClassMisc.asm
Comment (HubertHuckevoll) [reply Library/SpecUI/CommonUI/CWin/cwinClassMisc.asm:1]: Okay, I will revert this if it is only white-space changes.

File: Library/SpecUI/CommonUI/CWin/cwinUtils.asm
Comment (HubertHuckevoll) [reply Library/SpecUI/CommonUI/CWin/cwinUtils.asm:1]: Okay, I will revert this if it is only white-space changes.

File: Library/SpecUI/CommonUI/CWin/cwinFieldInit.asm (hunk 1 above)
Comment (HubertHuckevoll) [reply Library/SpecUI/CommonUI/CWin/cwinFieldInit.asm:536]: Yup, you're right, but the question is, what would be a good alternative?

File: Library/SpecUI/CommonUI/CWin/cwinFieldUncommon.asm (hunk 2 above)
Comment (HubertHuckevoll) [reply Library/SpecUI/CommonUI/CWin/cwinFieldUncommon.asm:1923]: I will translate this.

File: Library/SpecUI/CommonUI/CWin/cwinFieldUncommon.asm (hunk 3 above)
Comment (HubertHuckevoll) [reply Library/SpecUI/CommonUI/CWin/cwinFieldUncommon.asm:2526]: I will remove the comment.

File: Library/SpecUI/CommonUI/CWin/cwinFieldUncommon.asm (hunk 4 above)
Comment (HubertHuckevoll) [reply Library/SpecUI/CommonUI/CWin/cwinFieldUncommon.asm:2541]: I will remove the comment.

File: Library/Pref/ConfigUI/configui.asm [hunk 8]
```diff
@@ -281,19 +281,16 @@ PASS:		ds - object block
 RETURN:		ds - fixed up
 DESTROYED:	bx, cx, dx, di, bp
 
-%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%@	
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%@
 
 specificKey	char "specific",0
 
 ISUIDisableList	lptr \
-	EOGotoGeoManager,
```
[the remark this answers is not in the document]
Comment (HubertHuckevoll) [reply Library/Pref/ConfigUI/configui.asm:289]: The lists have changed a bit, see line 306 - 314...

File: Library/Pref/ConfigUI/configuiInterface.ui [hunk 9]
```diff
@@ -150,9 +151,18 @@ UIO3AutohideTaskbar = PrefBoolean {
...
+	hints = {
+		ATTR_GEN_INIT_FILE_CATEGORY { char "motifOptions",0 },
+		ATTR_GEN_INIT_FILE_KEY { char "taskBarMovable",0 }
+	}
+}
+
 UIO3RightClickHelp = PrefBoolean {
 	moniker = 'i', "Right Click Help in Documents";
-	identifier = 32;
+	identifier = 64;
```
[the remark this answers is not in the document]
Comment (HubertHuckevoll) [reply Library/Pref/ConfigUI/configuiInterface.ui:165]: It probably is not. However, I didn't know what identifier to assign and tried to make it consistent... looking at the identifiers in this file, which value(s) would you suggest?

File: Library/SpecUI/CommonUI/CSpec/cspecCUAS.ui (hunk 7 above)
[the remark this answers is not in the document]
Comment (HubertHuckevoll) [reply Library/SpecUI/CommonUI/CSpec/cspecCUAS.ui:1161]: Fixed.

File: Library/SpecUI/CommonUI/CWin/cwinFieldInit.asm (hunk 1 above)
Comment (HubertHuckevoll) [reply Library/SpecUI/CommonUI/CWin/cwinFieldInit.asm:536]: Okay, I've reworked the code to make the width of the E-Menu Button dependent on the field size (width). However, after many tries, I think the best solution for the height is to stay at 20 pixels as this creates the best look. I've tested several resolutions and it works pretty in-line with the rest of Motif.

File: Library/SpecUI/CommonUI/CWin/cwinBase.asm (hunk 5 above)
Comment (HubertHuckevoll) [reply Library/SpecUI/CommonUI/CWin/cwinBase.asm:1402]: It's a little bit unclear if this code is still needed, but I removed the condition and left it in, adding a comment.

File: Library/Pref/ConfigUI/configuiInterface.ui (hunk 9 above)
[the remark this answers is not in the document]
Comment (HubertHuckevoll) [reply Library/Pref/ConfigUI/configuiInterface.ui:165]: I fixed this.

File: Library/SpecUI/CommonUI/CWin/cwinFieldUncommon.asm (hunk 2 above)
Comment (HubertHuckevoll) [reply Library/SpecUI/CommonUI/CWin/cwinFieldUncommon.asm]: Reworked comments.

File: Library/User/UI/uiExpress.asm (hunk 6 above)
[the remark this answers is not in the document]
Comment (bluewaysw) [reply Library/User/UI/uiExpress.asm:678]: ok

File: Library/Pref/ConfigUI/configui.asm (hunk 8 above)
[the remark this answers is not in the document]
Comment (bluewaysw) [reply Library/Pref/ConfigUI/configui.asm:289]: ok

**Review by bluewaysw — APPROVED** [review]

**Review by bluewaysw — APPROVED** [review]

**Pull request merged**